composer/models/huggingface.py

```diff
@@ -272,20 +275,24 @@ def loss(self, outputs, batch):
             return outputs[0]
 
     def eval_forward(self, batch, outputs: Optional[Any] = None):
-        output = outputs if outputs else self.forward(batch)
-        if self.use_logits:
+        if 'eval_forward_handle' in batch:
```

Comment:
@dakinggg is this the canon way of doing this now? Or a temporary workaround until we have some way of getting the eval dataloader name to the `eval_forward` function?

The idea of datasets defining model forward functions, getting autocast right, getting distributed stuff right, makes me nervous. It also means the dataset's `eval_forward_handle` is now tied to HF models only right?

Reply:
My suggested approach right now is to define a custom collate function that packs a string "mode" into the batch, and then the `eval_forward` function uses that "mode" to determine which code path to run (this is what Sam did for his similar issue). I prefer this to actually packing the function reference into the batch. This deserves some docs that I have not written yet. As of right now, we don't have a plan to provide access to the `eval_dataloader.label` in `eval_forward`. I don't like the idea of changing the `eval_forward` function signature, but we could revisit if my suggestion above seems untenable.

As for your concerns about autocast, distributed, etc, I think it is ok because `eval_forward` gets called by composer inside all the correct contexts. Am I missing something here?